core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js

```diff
@@ -0,0 +1,165 @@
+                $scope.trainingData.splice(answerIndex, 1);
+              };
+
+              $scope.exit = function() {
+                $uibModalInstance.close();
+              };
+
+              $scope.submitAnswer = function(newAnswer) {
+                $scope.newAnswerIsResolved = false;
+                $scope.answerSuccessfullyAdded = false;
+
+                var interactionId = stateInteractionIdService.savedMemento;
+
+                var rulesServiceName =
+                  AngularNameService.getNameOfInteractionRulesService(
+                    interactionId);
+
+                // Inject RulesService dynamically.
+                var rulesService = $injector.get(rulesServiceName);
+
+                var newAnswerTemplate = (
+                  ExplorationHtmlFormatterService.getAnswerHtml(
+                    newAnswer, stateInteractionIdService.savedMemento,
+                    stateCustomizationArgsService.savedMemento));
+
+                var classificationResult = (
+                  AnswerClassificationService.getMatchingClassificationResult(
+                    _explorationId, _stateName, _state,
+                    newAnswer, rulesService));
+                var newAnswerFeedbackHtml = 'Nothing';
+                var newAnswerOutcomeDest = classificationResult.outcome.dest;
+                if (classificationResult.outcome.hasNonemptyFeedback()) {
+                  newAnswerFeedbackHtml = (
+                    classificationResult.outcome.feedback.getHtml());
+                }
+                if (newAnswerOutcomeDest === _stateName) {
+                  dest = '<em>(try again)</em>';
+                }
+
+                $scope.newAnswerTemplate = newAnswerTemplate;
+                $scope.newAnswerFeedbackHtml = newAnswerFeedbackHtml;
+                $scope.newAnswerOutcomeDest = newAnswerOutcomeDest;
+
+                var classificationType = (
+                  classificationResult.classificationCategorization);
+
+                // If answer is explicitly classified then show the
+                // classification results to the creator.
+                if (classificationType === EXPLICIT_CLASSIFICATION ||
+                    classificationType === TRAINING_DATA_CLASSIFICATION) {
+                  $scope.newAnswerIsResolved = true;
+                } else {
+                  TrainingDataService.trainAnswerGroup(
+                    answerGroupIndex, newAnswer);
+                  _rebuildTrainingData();
+                  $scope.answerSuccessfullyAdded = true;
+                }
+              };
+
+              $scope.openTrainUnresolvedAnswerModal = function(answerIndex) {
+                answer = $scope.trainingData[answerIndex].answer;
+                return TrainingModalService.openTrainUnresolvedAnswerModal(
+                  answer, true, $scope.finishTrainingAnswer);
+              };
+
+              $scope.finishTrainingAnswer = function() {
```

Comment:
Perhaps get rid of this and just have _rebuildTrainingData inline in line 152? Seems a bit odd to separate it out like this, IMO.

Reply:
I thought that this would be faster since we know which answer is being addressed. _rebuild would have to go through entire training data again.

Reply:
Not sure I understand, sorry. Line 156 just calls _rebuildTrainingData() directly. So how is this different then doing something like the following?
```
                return TrainingModalService.openTrainUnresolvedAnswerModal(
                  answer, true, _rebuildTrainingData);
```

Reply:
I don't have an objection to that, but that doesn't seem to be what $scope.finishTrainingAnswer() is currently doing at the moment either...

(The original change I proposed was meant to be a no-op in terms of functionality -- I was just talking about moving line 156 to line 152.)

Reply:
Oops, my bad. I thought we were talking about `finishTrainingAnswer()` method of teach_oppia_modal (the one you suggested to make anonymous function). This is a different case and your original comment stands right. Thanks.

Reply:
Done.